File: Week8/Assignment7/Statistics.ml
let mean lst =
  let total = List.fold_left (+) 0 lst in
  float_of_int total /. float_of_int (List.length lst)

let median lst =
  let sorted = List.sort compare lst in
  let n = List.length sorted in
  if n mod 2 = 1 then
    float_of_int (List.nth sorted (n / 2))
  else
    let mid1 = List.nth sorted (n / 2 - 1) in
    let mid2 = List.nth sorted (n / 2) in
    float_of_int (mid1 + mid2) /. 2.0

let mode lst =
  let sorted = List.sort compare lst in
  let rec find_mode lst current count max_val max_count =
    match lst with
    | [] -> max_val
    | h::t ->
      let new_count = if h = current then count + 1 else 1 in
      let (new_max_val, new_max_count) =
        if new_count > max_count then (h, new_count) else (max_val, max_count)
      in
      find_mode t h new_count new_max_val new_max_count
  in
  match sorted with
  | [] -> failwith "Empty list"
  | h::t -> find_mode t h 1 h 1

let () =
  let lst = [1; 2; 2; 3; 4; 5; 8] in
  Printf.printf "Mean: %.2f\n" (mean lst);
  Printf.printf "Median: %.2f\n" (median lst);
  Printf.printf "Mode: %d\n" (mode lst)

File: Week8/Assignment7/Statistics.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

double calculate_mean(int arr[], int n) {
    int sum = 0;
    for (int i = 0; i < n; i++)
        sum += arr[i];
    return (double)sum / n;
}

double calculate_median(int arr[], int n) {
    qsort(arr, n, sizeof(int), compare);
    if (n % 2 == 0)
        return (arr[n/2 - 1] + arr[n/2]) / 2.0;
    else
        return arr[n/2];
}

int calculate_mode(int arr[], int n) {
    qsort(arr, n, sizeof(int), compare);
    int mode = arr[0];
    int max_count = 1, count = 1;

    for (int i = 1; i < n; i++) {
        if (arr[i] == arr[i-1])
            count++;
        else
            count = 1;

        if (count > max_count) {
            max_count = count;
            mode = arr[i];
        }
    }
    return mode;
}

int main() {
    int arr[] = {1, 2, 2, 3, 4, 5, 8};
    int n = sizeof(arr)/sizeof(arr[0]);

    printf("Mean: %.2f\n", calculate_mean(arr, n));
    printf("Median: %.2f\n", calculate_median(arr, n));
    printf("Mode: %d\n", calculate_mode(arr, n));

    return 0;
}
